### DISCUSSION TITLE
Is it possible to hook field changes?

### DISCUSSION THREAD

**Discussion by kiber-io** [vfsfitvnm/frida-il2cpp-bridge#380] [Q&A] [answered]
For example, here is the energy field. 
![изображение](https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/60169133/6fefaa57-3658-4e7f-b1e7-5849d1e60b43)
It is used in different parts of the application as follows
![изображение](https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/60169133/00617945-5e6c-4e72-a574-4aa84500af79)
Hooking the getValue function is not an option, because it is a SafeInt function that is used in many fields. Can I somehow hook such an appeal to the field?

Or, there is also such a construction (I have no idea what it is called in C#)
![изображение](https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/60169133/4a2c578f-0fa4-4f2b-8a5e-93043cc9a696)
You could see the use of the Instance object in the screenshot above. Maybe you can hook it? But neither .method nor .field does not find it

**Comment by kiber-io** [comment, marked as answer]
I asked myself, I immediately found the answer myself :D

```c#
const csharp = Il2Cpp.domain.assembly('Assembly-CSharp').image

csharp.class('Account').method('get_Instance').implementation = function () {
    let acc = this.method('get_Instance').invoke()
    // acc is Account instance
    return acc
}
```
it turns out that for such a construction with Instance, its own get_\<field name\> method is created. Apparently, these are features of C#, and I practically didn't write on it.
Actually, in my case, it also gives the opportunity to replace the values of the fields at the same time

**Comment by kiber-io** [comment]
In general, if someone has an answer to the question is it possible to hook access to fields - I will be glad to feedback

**Comment by ekologic** [comment]
As a hack you get the in memory address of the SafeInt(handle) of the account instance and override the implementation of GetValue and inside check if the addresses matches then do what you want
I am doing something like this in one of my projects:
```
    SetActive.implementation = function (value: boolean) {
       globalData[this.handle.toString()] =
       globalData[this.handle.toString()] || {};
       globalData[this.handle.toString()].IsActive = value;

      this.method("SetActive").invoke(value);
    };
```

But yeah if there is other clean way of doing it, would be happy to hear other suggestions 😄

**Comment by ekologic** [comment]
> Or, there is also such a construction (I have no idea what it is called in C#) 

It is a design pattern and it is called singleton and it is not c# specific, but in general to OOP

**Reply by kiber-io** [reply]
I know what a singleton is and what such a construction is. but in different languages there are different ways to create a singleton, and in C#, as we can see, it has its own.  in the source code, something similar to the code from the screenshot is set, and under the hood, apparently, the get_Intance method is created, which is called as a result.
